File: A429/A429BasicFDX/FDXClient_C++/FDXDispatcher.cpp
/*-----------------------------------------------------------------------------
Module: CANoeFDX
Interfaces: FDXDispatcher
-------------------------------------------------------------------------------
Datagram dispatcher class for CANoe FDX protocol
-------------------------------------------------------------------------------
Copyright (c) Vector Informatik GmbH. All rights reserved.
-----------------------------------------------------------------------------*/
#include "stdafx.h"
#include "FDXDispatcher.h"



FDXDispatcher::FDXDispatcher() 
 : mNextExpectedSeqNr (CANoeFDX::kSequenceNumberSessionStart)
{}



void FDXDispatcher::DispatchDatagram(FDXDatagram& datagram)
{
  //
  // A) Check datagram header
  //
  CANoeFDX::DatagramHeader* header = reinterpret_cast<CANoeFDX::DatagramHeader*>(datagram.Buffer());
  {
    if (datagram.Size()<sizeof(CANoeFDX::DatagramHeader))
    {
      OnFormatError(std::string("Size of datagram is too small"));
      return;  // datagram is smaller as a header
    }

    if (header->fdxSignature != CANoeFDX::kFdxSignature )
    {
      OnFormatError(std::string("Signature doesn't match"));
      return;  // different value for magic cookie 
    }

    if (header->fdxMajorVersion != CANoeFDX::kFdxMajorVersion) // major version does not fit
    {
      OnFormatError(std::string("Major FDX Version doesn't match"));
      return; // version error
    }

    // Check Sequence Number
    if (header->sequenceNumber == CANoeFDX::kSequenceNumberUnused)
    {
       // sequence numbering unused => check nothing
    }
    else
    {
      if ( (header->sequenceNumber & 0x7FFF) != mNextExpectedSeqNr)
      {
        OnSequenceNumberError(header, mNextExpectedSeqNr);
      }
      if (header->sequenceNumber & CANoeFDX::kSequenceNumberSessionEndFlag)
      {
        mNextExpectedSeqNr = CANoeFDX::kSequenceNumberSessionStart;
      }
      else
      {
        mNextExpectedSeqNr = IncrementSequenceNumber(header->sequenceNumber);
      }
    }
  }


  //
  // B) Check The memory structure of the following commands.
  //    Do this check for the whole datagram before any request is processed
  {
    DWORD offset = sizeof(CANoeFDX::DatagramHeader);
    DWORD remainingBytes = datagram.Size() - offset;

    for (DWORD i=0; i < header->numberOfCommands; i++)
    {
      if (remainingBytes<sizeof(CANoeFDX::CommandHeader))
      {
        OnFormatError(std::string("command header is too small"));
        return;  // too small for command header
      }

      CANoeFDX::CommandHeader* command = reinterpret_cast<CANoeFDX::CommandHeader*>((char*)datagram.Buffer()+offset);
      if (remainingBytes<command->commandSize)
      {
        OnFormatError(std::string("datagram is too small for command"));
        return;  // datagram is to small for this command
      }

      offset += command->commandSize;
      remainingBytes -= command->commandSize;
    }

    if (remainingBytes != 0)
    {
      OnFormatError(std::string("unused data at the end of the datagram"));
      return;  // there are some unused data bytes at the end of the datagram
    }
  }


  //
  // C) Check the internal structure of the commands
  //
  {
    DWORD offset = sizeof(CANoeFDX::DatagramHeader);
    for (DWORD i=0; i < header->numberOfCommands; i++)
    {
      CANoeFDX::CommandHeader* command = reinterpret_cast<CANoeFDX::CommandHeader*>((char*)datagram.Buffer()+offset);
      switch(command->commandCode)
      {
      case CANoeFDX::kCommandCode_DataExchange:
        if (command->commandSize < CANoeFDX::kDataExchangeBaseSize)
        {
          OnFormatError(std::string("data exchange command is too small"));
          return;  // invalid size for DataExchange command, command is to small
        }
        if (command->commandSize < CANoeFDX::kDataExchangeBaseSize+ static_cast<CANoeFDX::DataExchangeCommand*>(command)->dataSize )
        {
          OnFormatError(std::string("invalid size for data exchange command"));
          return;  // invalid size for DataExchange command, payload does not fit into command
        }
        break;

      case CANoeFDX::kCommandCode_Status:
        if (command->commandSize<sizeof(CANoeFDX::StatusCommand))
        {
          OnFormatError(std::string("status command is too small"));
          return;
        }

      case CANoeFDX::kCommandCode_DataError:
        if (command->commandSize<sizeof(CANoeFDX::DataErrorCommand))
        {
          OnFormatError(std::string("data error command is too small"));
          return;
        }
                
      case CANoeFDX::kCommandCode_HardwareChanged:
        if (command->commandSize<sizeof(CANoeFDX::CommandHeader))
        {
          OnFormatError(std::string("hardware changed command is too small"));
          return;
        }
        break;

      default:
        // ignore any unknown command
        break;
      }
      offset += command->commandSize;
    }
  }

  //
  // Dispatch command
  //
  {
    DWORD offset = sizeof(CANoeFDX::DatagramHeader);
    for (DWORD i=0; i < header->numberOfCommands; i++)
    {
      CANoeFDX::CommandHeader* command = reinterpret_cast<CANoeFDX::CommandHeader*>((char*)datagram.Buffer()+offset);
      switch(command->commandCode)
      {
      case CANoeFDX::kCommandCode_DataExchange:
        OnDataExchange(header, static_cast<CANoeFDX::DataExchangeCommand*>(command));
        break;
      case CANoeFDX::kCommandCode_Status:
        OnStatus(header, static_cast<CANoeFDX::StatusCommand*>(command));
        break;
      case CANoeFDX::kCommandCode_DataError:
        OnDataError(header, static_cast<CANoeFDX::DataErrorCommand*>(command));
        break;
      default:
        // ignore any other command
        break;
      }
      offset += command->commandSize;
    }
  }
}

void FDXDispatcher::OnFormatError(std::string &error)
{
  printf("Format Error: %s\n", error);
}

void FDXDispatcher::OnSequenceNumberError(CANoeFDX::DatagramHeader* header, CANoeFDX::uint16 expectedSeqNr)
{
  printf("Sequence Number Error: expected SeqNr: %d, received SeqNr: %d\n", expectedSeqNr, header->sequenceNumber);
}

void FDXDispatcher::OnStatus(CANoeFDX::DatagramHeader* header, CANoeFDX::StatusCommand* command)
{
  if (command->measurementState == CANoeFDX::kMeasurementState_NotRunning)
  {
    printf("CANoe measurement is not running\n");
  }
}

void FDXDispatcher::OnDataError(CANoeFDX::DatagramHeader* header, CANoeFDX::DataErrorCommand* command)
{
  if (command->dataErrorCode != CANoeFDX::kDataErrorCode_MeasurementNotRunning)
  {
    printf("Data error for group ID %d, error code %d\n", command->groupID, command->dataErrorCode);
  }
}

File: A429/A429BasicFDX/FDXClient_C#/FDXDispatcher.cs
/*-----------------------------------------------------------------------------
Module: CANoeFDX
Interfaces: FDXDispatcher
-------------------------------------------------------------------------------
Datagram dispatcher class for CANoe FDX protocol
-------------------------------------------------------------------------------
Copyright (c) Vector Informatik GmbH. All rights reserved.
-----------------------------------------------------------------------------*/
namespace FDXClient
{
  using System;

  /// <summary>
  /// This class implements a Dispatcher for incoming datagrams and implements
  /// basic handlers for standard behavior.
  /// </summary>
  /// <remarks>
  /// To handle exchange data you need to inherit from FDXDispatcher and implement 
  /// your own data handler.
  /// </remarks>
  public class FDXDispatcher
  {
    public FDXDispatcher()
    {
      this.mNextExpectedSequenceNumber = FDXHelperSequenceNumber.kSequenceNumberStart;
    }

    public void DispatchDatagram(ref FDXDatagram datagram)
    {
      var lCopy = datagram.Buffer;

      //if the received bytes are too short, skip
      if (lCopy.Length < 16)
      {
        this.OnFormatError("Datagram too short");
        return;
      }

      // extract datagram header information
      var header = DatagramHeader.DeSerialize(lCopy);

      //if the signature doesn't match, skip
      if (header.FDXSignature != FDXHelper.kFDXSignature)
      {
        this.OnFormatError("Signature mismatch");
        return;
      }

      //if the major version doesn't match, skip
      if (header.FDXMajorVersion != FDXHelper.kFDXMajorVersion)
      {
        this.OnFormatError("Incorrect Major version");
        return;
      }

      //if the minor version doesn't match, skip
      if (header.FDXMinorVersion != FDXHelper.kFDXMinorVersion)
      {
        this.OnFormatError("Incorrect Minor Version");
        return;
      }

      //checking the sequence number
      if (header.SequenceNumber == FDXHelperSequenceNumber.kSequenceNumberUnused)
      {
        //we don't use sequence numbering
      }
      else
      {
        // check if the received sequence number is the expected
        if ( (header.SequenceNumber & 0x7FFF) != this.mNextExpectedSequenceNumber)
        {
          this.OnSequenceError(this.mNextExpectedSequenceNumber, header.SequenceNumber);
        }
        // if the session end sequence number arrives, reset the expected sequence number
        else if ((header.SequenceNumber & FDXHelperSequenceNumber.kSequenceNumberSessionEndFlag) != 0)
        {
          this.mNextExpectedSequenceNumber = FDXHelperSequenceNumber.kSequenceNumberStart;
        }
        else
        {
          // if expected sequence number matches the received sequence number
          this.mNextExpectedSequenceNumber = FDXDatagram.IncrementSequenceNumber(header.SequenceNumber);
        }
      }

      //extract the number of commands and set the position
      var numOfCommands = header.NumberOfCommands;
      var commandPosAndLength = DatagramHeader.Size();

      //create a buffer for the single commands
      var data = new byte[lCopy.Length - commandPosAndLength];

      for (var i = 0; i < numOfCommands; ++i)
      {
        //copy command number i into the buffer
        Array.Copy(lCopy, commandPosAndLength, data, 0, lCopy.Length - commandPosAndLength);

        //extract the header information and copy the data from the buffer into the concrete command
        // buffer
        var cmdHead = CommandHeader.DeSerialize(data);
        var concreteCommand = new byte[cmdHead.CommandSize];
        Array.Copy(data, 0, concreteCommand, 0, concreteCommand.Length);
        commandPosAndLength += concreteCommand.Length;

        // with the command code we can decide what to do with the received command
        switch (cmdHead.CommandCode)
        {
          case FDXHelperCommandCode.kCommandCodeDataExchange:
            {
              var cmd = DataExchangeCommand.DeSerialize(concreteCommand);
              this.OnDataExchange(cmd);
              break;
            }
          case FDXHelperCommandCode.kCommandCodeStatus:
            {
              var cmd = StatusCommand.DeSerialize(concreteCommand);
              this.OnStatus(cmd);
              break;
            }
          case FDXHelperCommandCode.kCommandCodeSequenceNumberError:
            {
              var com = SequenceNumberErrorCommand.DeSerialize(concreteCommand);
              this.OnSequenceError(com.ExpectedSeqNr, com.ReceivedSeqNr);
              break;
            }
        }

      }
    }

    // Data handler for format error, prints out the error message
    public virtual void OnFormatError(string errorString)
    {
      Console.Out.WriteLine("Format Error: {0}", errorString);
    }

    // Data handler for status, prints out the state of the measurement and the state
    public virtual void OnStatus(StatusCommand stData)
    {
      var elapsedTime = new TimeSpan(stData.Timestamp);
      Console.Out.WriteLine("MeasurementState: {0} TimeStamp: {1}", stData.MeasurementState, elapsedTime);
    }
    
    // Data handler for Sequence numbers, prints out the expected and the received sequence number
    public virtual void OnSequenceError(UInt16 expected, UInt16 received)
    {
      Console.Out.WriteLine("Sequence Number Error: Expected Seq Nr: {0}, Received Seq Nr: {1}", expected, received);
    }

    // Data handler for data errors, prints out the groupId and error code
    public virtual void OnDataError(DataErrorCommand error)
    {
      Console.Out.WriteLine("Data Error: GroupId: {0}, Errorcode: {1}", error.GroupId, error.DataErrorCode);
    }

    //Empty data handler for data exchange, can be overwritten by an own FDXDispatcher that only implement this function
    public virtual void OnDataExchange(DataExchangeCommand exData) { }

    #region Variables
    private UInt16 mNextExpectedSequenceNumber;
    #endregion
  }
}
